### DISCUSSION TITLE
Sending child actions directly from the parent to update the child

### DISCUSSION THREAD

**Discussion by iwt-sebastian-boldt** [pointfreeco/swift-composable-architecture#1952] [Q&A] [answered]
I have a brief question regarding parent-child relationships and a possible method of updating the child from the parent. 
Suppose I have a parent feature that encapsulates a child in the typical TCA manner. 
This involves defining a child property in the state and a case in the action enum that is scoped to the correct reducer. 
The child reducer is responsible for updating the text if it receives the update action. 
Additionally, the child reducer may have some proprietary business logic that it does not want to expose to external parties. Therefore, updating the text should be the child's responsibility.

```swift 

struct Parent: ReducerProtocol {
    struct State: Equatable {
        var child: Child.State
    }

    enum Action: Equatable {
        case child(Child.Action)
    }

    var body: some ReducerProtocol<State,Action> {
        Scope(state: \.child, action: /Action.child) {
            Child()
        }
        Reduce { state, action in
            return .none
        }
    }
}

struct Child: ReducerProtocol {
    struct State: Equatable {
        var text: String
    }
    enum Action: Equatable {
        case update
    }

    var body: some ReducerProtocol<State,Action> {
        Reduce { state, action in
            switch action {
                case .update:
                state.text = // do validation or api calls etc. that is related to the Childs Domain
                return .none
            }
        }
    }
}
```

And in my Parent's View, I want to have a Button that should trigger an action in the child.
Technically, I would be able to send the action directly to the child from the view 
or from the reducer by using `Effect(value: .child(.update))`

```swift
struct ParentView: View {
    let store: StoreOf<Parent>

    var body: some View {
        WithViewStore(self.store) { viewStore in
            Button(action: {
                viewStore.send(.child(.update))
                // or..
                viewStore.send(.didPressButton) // in the reducer return Effect(value: .child(.update))
            }, label: {
                Text("Trigger Child Action")
            })
        }
    }
}
```

Now my question is, is this a common pattern to trigger updates in child views?
Or does it violate some of the principles of TCA?
I am not sure because it works but it feels kinda wrong

**Comment by mbrandonw** [comment, marked as answer]
Hi @iwt-sebastian-boldt, neither of the styles you have sketched out is recommended in TCA.

First, actions should be named after what the user did in the UI, not what you want to do inside the reducer. So the action `.didPressButton` is clear and concise because the user just did something in the UI. However, `.update` is opaque and no one unfamiliar with the intricacies of the child domain will know when that action should be sent.

Further, you should not think of actions as just weird kinds of methods defined for your domain. They are not methods, but instead a concrete description of things that can happen in the UI, and sending actions is a lot more expensive than calling methods. We have an [article](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/performance#Sharing-logic-with-actions) that goes into detail in this topic.

My suggestion for your set up is for the child state to expose a mutating helper method that can do your logic, including returning effects:

```swift
extension Child.State {
  mutating func update(…) -> Effect<Child.Action> {
    …
  }
}
```

Note that because this is a _method_ and not an _action_, it _is_ appropriate to name it after  what it does rather than what the user is doing. And because this is a simple method you can call it from anywhere, including the child or parent domains.

For example, in the parent you could do the following:

```swift
case .buttonTapped:
  return state.child.update().map(Action.child)
```

That will run the mutating `update` method on the child state and return some child effects, which you then need to bundle back up into the parent domain using `.map(Action.child)`.

This style gives you maximum flexibility since you can call `child.update()` directly anywhere you have access to child state, _and_ it is most performant because you are just invoking a simple method and not sending a whole new action into the system.

**Reply by lukeredpath** [reply]
@mbrandonw interesting idea to have mutating functions on `State` that return effects. I've usually put my reusable effects into helper functions on the _reducer_ itself although if it also needs to mutate state that means passing along some `inout State` too.

The only weird thing about this is that if the effect needs a dependency (which it probably does), you'd need to add the `@Dependency` properties to your state instead of your reducer. Does this seem like a code smell?

**Reply by jgoodrick** [reply]
You don't really need to add them to the state, right? Just to the method you're defining? Since you can call them inline within the method body? (Which can be defined in whatever module needs to use it)

**Reply by mbrandonw** [reply]
> I've usually put my reusable effects into helper functions on the _reducer_ itself although if it also needs to mutate state that means passing along some `inout State` too.

@lukeredpath I think that style works great when calling shared logic directly in the same reducer, but it breaks down when wanting to invoke shared logic in a _different_ reducer. That's where the mutating method style shines.

> The only weird thing about this is that if the effect needs a dependency (which it probably does), you'd need to add the @Dependency properties to your state instead of your reducer. Does this seem like a code smell?

You can declare the dependency directly in the mutating method:

```swift
extension Child.State {
  mutating func update(…) -> Effect<Child.Action> {
    @Dependency(\.apiClient) var apiClient
    …
  }
}
```

As long as `update()` is only ever invoked within a reducer context it will pick up the correct dependencies just fine. If you call this method outside of a reducer, then the dependencies will go back to their defaults. You could also pass the dependency to the `update` method if you want to be super explicit just for this method.

**Reply by lukeredpath** [reply]
That makes sense. I do actually like this approach for sharing logic from the child to the parent.

Another approach I've taken where I've had some complex effect logic I want to share across multiple domains is to have a high-level client that I can use as a dependency anywhere - this client is probably implemented internally using other dependencies but wraps up more complex effects into simple `async` functions that I can call from any reducer.

In fact, I've just refactored part of our app logic from an older style that used the reducer to implement the complex logic (this was pre-swift concurrency) and requiring the parent domain to fire a child action to start some effect to this approach. 

Before, to refresh session data for our logged in user (which required firing off a lot of different API requests) we'd have a high-level reducer to manage all of the effects and state and you'd need to do something like:

```swift
case .didEnterForeground:
   return .send(.child(.refreshSessionData))
case let .child(.refreshFinished(sessionData)):
  state.sessionData = SessionData
  return .none
```

Now we have a `SessionClient` that just has a `refreshSessionData: () async throws -> SessionData` function that we can call:

```swift
@Dependency(\.sessionClient) private var sessionClient

case .didEnterForeground:
  return .task { 
    await .refreshedSessionData(
      TaskResult { try await sessionClient.refreshSessionData() }
    }
  }
```

Not only is this much easier to deal with, but it gives us two options for how we approach testing this:

1. Integration style, where we use the live session client but mock out its internal dependencies - as this tests the internal logic of the live session client this only really needs to be done in a few focused integration tests.
2. Non-integrated, where we can just mock the session client itself (replacing a dozen API client request mocks with a single mock of the `refreshSessionData` function). We can do this everywhere else where we are only concerned with testing the domain itself.

**Reply by tomassliz** [reply]
@lukeredpath What's the difference between yours approach (I'm using it too) and the one introduced on the beginning of this discussion by @iwt-sebastian-boldt? @mbrandonw recommends using a state mutating function defined on child states because it's more efficient.

Now I'm confused - when is it ok to call actions on child domain from the parent?

**Reply by lukeredpath** [reply]
@tomassliz I'm not calling actions on the child domain - the post above is an example of how I've refactored some of our code to *stop* calling child actions from the parent by moving some reusable behaviour into a client that we access through the dependency system. The way we _used_ to do it is closer to the original post.

**Reply by tomassliz** [reply]
@lukeredpath Oh, now I get it, sorry for misunderstanding.

**Reply by vvisionnn** [reply]
@mbrandonw really thanks for the response but I'm still little confused about the action and the method, here is what I have learned,
if I just want to update the child state from the parent only, and don't want this action fired into the action system, I should use the mutating method that defined for child state
what if I want to modify the child state and notify all the parent feature**s** (not only the child's parent feature, maybe the upper level like the parent feature of the child's parent), should I use the action trigger the child state update?

**Reply by mbrandonw** [reply]
Hi @vvisionnn, communication from child _to_ parent is a great use case for actions.

We even discussed this in one of our [episodes](https://www.pointfree.co/episodes/ep222-composable-navigation-tabs) (at [42:11](https://www.pointfree.co/episodes/ep222-composable-navigation-tabs#t2531)), and Krzysztof Zabłocki also [wrote](https://www.merowing.info/boundries-in-tca/) about this pattern. [Here](https://github.com/search?q=repo%3Apointfreeco%2Fisowords+DelegateAction&type=code) are also some examples of it in our isowords codebase.

Dependencies are another way of doing this if you need to notify _many_ parent features of changes. That may be more ergonomic than each parent feature destructuring child actions in order to figure out what is going on.

**Reply by acosmicflamingo** [reply]
@mbrandonw this breaks down when the dependency within child state's mutating function requires being run in main thread, right?

I was thinking of doing this:
```swift
extension SplitScreen.State {
  @MainActor
  mutating func updateOrientation() -> Effect<SplitScreen.Action> {
    @Dependency(\.deviceClient) var deviceClient: DeviceClient
    orientation = deviceClient.interfaceOrientation
    return .none
  }
}
```
However, I realize that `deviceClient.interfaceOrientation` has a dependency on UIApplication (DependencyAdditions' ApplicationDependency actually), whose properties need to be retrieved from main thread. Although the fix for that would be to run it within an Effect closure, that would mean I am now attempting to modify state within a concurrent context, which is the biggest no-no ;)

Edit: I'd like to think I can just slap on @MainActor to my SplitScreen reducer since it should in fact force everything to be run in main thread anyway, but I then see this warning: `Main actor-isolated property 'body' cannot be used to satisfy nonisolated protocol requirement`

**Reply by tomassliz** [reply]
@mbrandonw the child state mutating function has a little downside - you can't call shared logic in it because it's defined in the feature scope.

```swift
struct Child: ReducerProtocol {
  struct State {
    mutating func update(…) -> Effect<Child.Action> {
      …
    }
  }


  func sharedComputation(state: inout State) -> EffectTask<Action> {
    // Some shared work to compute something.
    return .run { send in ... }
  }
}
```
I wanted to mutate child states directly but this seems like I don't have any option to avoid parent -> child action calls in this case 😞

**Reply by lukeredpath** [reply]
@tomassliz the solution to this is to simply move the shared helpers out of the feature scope - you could move them into the file scope and mark them as fileprivate.

Also, in your example your shared helper takes an `inout State` which suggests you could invert the dependency and move that helper on to your child state too - you could also mark this as `private` if you don't want it to be called from outside the feature:

```swift
struct State {
  private mutating func sharedComputation() -> EffectTask<Action> {
    // stuff
  }
}
```

Now you can call it from both your public state function and your child reducer.

**Reply by tomassliz** [reply]
@lukeredpath I ended with a similar solution. But I'm still wondering if having shared helpers within the child state scope is a good idea. It will work but it feels weird to me.

**Reply by lukeredpath** [reply]
It feels perfectly natural and functional to me for a value type to have a mutating function that returns an effect. You can use normal access control to control who can call it.

**Reply by mariano-mix** [reply]
> ```swift
> case .buttonTapped:
>   return state.child.update().map(Action.child)
> ```
Thanks for sharing this. One question regarding it, is that in this case you are also returning an effect that could potentially be a child action that feeds the system if let's say the child's update method returned some effects of its own. So basically the gain here is that for the update we do it directly to the state without feeding an extra action to the system which is costly, am I getting this right?

**Reply by lukeredpath** [reply]
Yes, it renders the intermediate send redundant.

**Reply by jerryga** [reply]
In the mutating method, we can modify the data in the child. But how can we call a child's action directly?
```swift
extension Child.State {
  mutating func playCurrent() -> Effect<Child.Action>  {
      self.index = self.index + 1
      return .run { send in
          await send(.play) // **_### this code is not called. why?_**
      }.cancellable(id: SubtitleCore.CancelID.play, cancelInFlight: false)
  }
}
```

**Reply by lukeredpath** [reply]
If you're calling this method from a parent reducer, you need to make sure you're returning the effect (you'll need to call `.map` to map the child action to an action in the parent) e.g.:

```swift
case .someParentAction:
  return state.child.playCurrent().map(Action.child)
```

Or, if you're already performing some other effect in the parent:

```swift
case .someParentAction:
  let childEffect = state.playCurrent()
  return .merge(
    .run { send in ... },
    childEffect.map(Action.child)
  )
```

**Reply by jerryga** [reply]
Hi, lukeredpath, please check the playCurrent() function. In that function, the following code is not called when I set a breakpoint on it.
".play" is an action of the child. In the reducer of the child, ".play" is not triggered.

**Reply by lukeredpath** [reply]
@jerryga I'm not sure what more I can do to help you here - how are you calling `playCurrent()` and from where? Is your child reducer composed into the parent reducer?

**Reply by kawynfurstoss** [reply]
@mbrandonw how would you do this using the @Reducer macro? I get a compiler error when trying to make an extension for Reducer.State when the Reducer struct uses the @Reducer macro

**Reply by kevinanderson7** [reply]
@kawynfurstoss you would just either move the extension to a separate file or instead of declaring the mutating func as part of an extension, you could put it directly in the Child.State declaration. See the "Circular Reference Errors" section here for more information: https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/reducers#Gotchas

**Reply by arkulpaKonstantin** [reply]
I discovered this thread today and really like the pattern of mutating the child state directly instead of sending actions to the child reducer. In my case I can now get rid of several actions that were only there for that case but I run into issues with testing. 

How can I setup a test for the child reducer without these actions? Do I need to add these actions back just for the sake of testing?

**Comment by lukeredpath** [comment]
If something has to be run on an actor (any actor) then it needs to be
an effect but you should be able to synchronously read the value from
your device client in your mutating state function because the TCA store
runs on the main thread anyway. 

On March 5, 2023, GitHub ***@***.***> wrote:
> @mbrandonw <https://github.com/mbrandonw> this breaks down when the
> dependency within child state's mutating function requires being run
> in main thread, right?
>
> I was thinking of doing this:
>
> extension SplitScreen.State { @MainActor mutating func
> updateOrientation() -> Effect<SplitScreen.Action> {
> @Dependency(\.deviceClient) var deviceClient: DeviceClient orientation
> = deviceClient.interfaceOrientation return .none } }
> However, I realize that deviceClient.interfaceOrientation has a
> dependency on UIApplication (DependencyAdditions'
> ApplicationDependency actually), whose properties need to be retrieved
> from main thread. Although the fix for that would be to run it within
> an Effect closure, that would mean I am now attempting to modify state
> within a concurrent context, which is the biggest no-no ;)
>

—
Reply to this email directly, view it on GitHub
<https://github.com/pointfreeco/swift-composable-
architecture/discussions/1952#discussioncomment-5209262>, or unsubscribe
<https://github.com/notifications/unsubscribe-
auth/AAAAEZJTFU2RSSNSIBHBQUTW2T6JLANCNFSM6AAAAAAVLWLZP4>.
You are receiving this because you were mentioned.Message ID:
<pointfreeco/swift-composable-architecture/repo-
***@***.***>

**Reply by acosmicflamingo** [reply]
Gotcha. I now see that this error actually has to do with how I designed the dependency. Because interfaceOrientation is being retrieved from UIApplication and that needs to be extracted from the main thread, I must add @MainActor to property like this:

```swift
extension DeviceClient {
  public var idiom: UIUserInterfaceIdiom {
    _implementation.idiom
  }

  @MainActor
  public var interfaceOrientation: UIInterfaceOrientation {
    @Dependency(\.application) var application
    return _implementation.interfaceOrientation(application)
  }
```

That means that in my example, I must add @MainActor here (although it leads to more problems):
```swift
extension SplitScreen.State {
  @MainActor
  public mutating func updateOrientation() -> Effect<SplitScreen.Action> {
    @Dependency(\.deviceClient.interfaceOrientation) var interfaceOrientation:
      UIInterfaceOrientation
    orientation = interfaceOrientation
    return .none
  }
}
```

I found the fix though that works perfectly :)

```swift
extension SplitScreen.State {
  public mutating func updateOrientation() -> Effect<SplitScreen.Action> {
    @Dependency(\.deviceClient.interfaceOrientation) var interfaceOrientation: UIInterfaceOrientation
    orientation = interfaceOrientation
    return .none
  }
}
```

**Reply by lukeredpath** [reply]
@acosmicflamingo if you're returning a `.none` effect you don't really need to return anything from this function - it can just be a normal mutating function with no return value.

**Reply by acosmicflamingo** [reply]
@lukeredpath true. That would certainly remove boilerplate code for the parent reducers making the calls. Thanks for the tip!

**Comment by WHuangz88** [comment]
> In fact, I've just refactored part of our app logic from an older style that used the reducer to implement the complex logic (this was pre-swift concurrency) and requiring the parent domain to fire a child action to start some effect to this approach.

Before, to refresh session data for our logged in user (which required firing off a lot of different API requests) we'd have a high-level reducer to manage all of the effects and state and you'd need to do something like:


Hi @lukeredpath , I m still quite confuse with the solution for this statement, how do you pass the information from parent to the child if the case is like pulling to refresh that only happened in the parent domain but all the children need to have the information in order to update all the dependency inside child reducer?

> First, actions should be named after what the user did in the UI, not what you want to do inside the reducer. So the action .didPressButton is clear and concise because the user just did something in the UI. However, .update is opaque and no one unfamiliar with the intricacies of the child domain will know when that action should be sent.


and for what @mbrandonw said, I totally agreed because only the child know what happened to its components, thus calling the action from parent domain doesn't make sense, but it's a different case with pulling to refresh right? where all child domains rely on the parent.

**Comment by idelfonsog2** [comment]
Insights: Following partially Brandon's suggestion, I was trying to send one of the child actions in this mutating method, but the child state was not changing. So I did the inverse; I added the reducers mutating logic in this `update()` method and called it from my child reducer `state.update()`.

I also had to update the parent state after calling update()

```swift
_ = item.update()
state.items.updateOrAppend(item)
```

Hope it helps

**Comment by iwt-sebastian-boldt** [comment]
It seems that the docs were updated and there is a new preferred way to do this 

```swift 
case .buttonTapped:
  return reduce(into: &state, action: .child(.refresh))
 ```
 
https://swiftpackageindex.com/pointfreeco/swift-composable-architecture/main/documentation/composablearchitecture/performance#Sharing-logic-in-child-features

**Reply by bradhowes** [reply]
FYI, for anyone who gets here: be aware that the code in the previous comment  is no longer recommended, and the doc has been updated to just show the less performant `send(.child(.refresh))`. Now with TCA 1.25 calling `reduce(into:action:)` generates a warning with guidance on how to fix.
